File: ts/src/hooks/hosts/gemini.ts
import { loadCapabilities, level, hostEvent as capHostEvent, type Capabilities } from "../../capabilities";
import type { Event } from "../../events";
import { get } from "../../registry";
import { UnsupportedActionError, UnsupportedEventError, SchemaError } from "../../errors";
import type { Codec } from "../codec";
import type { Input } from "../input";
import type { Decision } from "../decision";

/** The Gemini CLI codec. */
class GeminiCodec implements Codec {
  private readonly caps: Capabilities;
  private readonly hostToCanonical: Map<string, Event>;

  constructor() {
    this.caps = loadCapabilities("gemini");
    this.hostToCanonical = decodeMap(this.caps);
  }

  host(): string {
    return "gemini";
  }

  capabilities(): Capabilities {
    return this.caps;
  }

  // encodeInput fails with UnsupportedEventError when the event has no
  // native or close mapping: synthesized, unsupported, or unclassified
  // events cannot be encoded as a Gemini-native payload since Gemini
  // itself never emits them.
  encodeInput(input: Input): unknown {
    requireNativeOrClose(this.caps, "gemini", input.event);
    const hostEventName = capHostEvent(this.caps, input.event);
    if (hostEventName === undefined) {
      throw new UnsupportedEventError("gemini", input.event);
    }

    const out: Record<string, unknown> = { ...(input.extra ?? {}) };
    out.hook_event_name = hostEventName;
    out.session_id = input.sessionId;
    out.cwd = input.cwd;
    switch (input.event) {
      case "PreToolUse":
        out.tool_name = input.toolName;
        out.tool_input = input.toolInput;
        break;
      case "PostToolUse":
        out.tool_name = input.toolName;
        out.tool_input = input.toolInput;
        out.tool_response = input.toolResponse;
        break;
      case "UserPromptSubmit":
        out.prompt = input.prompt;
        break;
      default:
        break;
    }
    return out;
  }

  // decodeInput looks up the canonical event in the native-wins decode
  // map on hook_event_name; a name not there is SchemaError, never a
  // silent fallback.
  decodeInput(raw: unknown): Input {
    const m = asRecord(raw, "gemini input");
    const hostEventName = m.hook_event_name;
    if (typeof hostEventName !== "string" || hostEventName === "") {
      throw new SchemaError("gemini input: missing hook_event_name");
    }
    const event = this.hostToCanonical.get(hostEventName);
    if (event === undefined) {
      throw new UnsupportedEventError("gemini", hostEventName);
    }

    const known = new Set(["hook_event_name", "session_id", "cwd", "tool_name", "tool_input", "tool_response", "prompt"]);
    const extra: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(m)) {
      if (!known.has(k)) extra[k] = v;
    }

    const input: Input = {
      event,
      sessionId: typeof m.session_id === "string" ? m.session_id : "",
      cwd: typeof m.cwd === "string" ? m.cwd : "",
      extra,
    };
    if (typeof m.tool_name === "string") input.toolName = m.tool_name;
    if (isRecord(m.tool_input)) input.toolInput = m.tool_input;
    if (isRecord(m.tool_response)) input.toolResponse = m.tool_response;
    if (typeof m.prompt === "string") input.prompt = m.prompt;
    return input;
  }

  // encodeDecision emits Gemini's native decision envelope. Gemini has no
  // native "warn"; nerv folds warn into decision "allow" and, when the
  // message is non-empty, carries it in BOTH "reason" and "systemMessage"
  // (invariant 2: never mapped to block). Gemini does not vary its wire
  // shape by event (invariant 6): the event gate below only rejects an
  // event Gemini cannot express at all; it does not change the shape.
  encodeDecision(d: Decision): { stdout: unknown; exit: number } {
    this.checkEvent(d.event);
    const codes = hostExitCodes(this.caps);
    switch (d.action) {
      case "allow":
        return { stdout: { decision: "allow" }, exit: codes.allow };
      case "warn": {
        const out: Record<string, unknown> = { decision: "allow" };
        if (d.message) {
          out.reason = d.message;
          out.systemMessage = d.message;
        }
        // host.yaml carries no warn exit code for gemini; warn exits like allow.
        return { stdout: out, exit: codes.allow };
      }
      case "rewrite":
        // host.yaml carries no rewrite exit code for gemini; rewrite
        // still allows the tool call to proceed, so it exits like allow.
        return {
          stdout: { decision: "allow", hookSpecificOutput: { tool_input: d.rewrite } },
          exit: codes.allow,
        };
      case "block":
        return { stdout: { decision: "deny", reason: d.message ?? "" }, exit: codes.block };
      default:
        throw new UnsupportedActionError(`gemini cannot express action ${String(d.action)}`);
    }
  }

  // decodeDecision reads Gemini's native decision envelope. Empty stdout,
  // or JSON that carries none of Gemini's known decision keys, both fall
  // back to actionForExit(exit): the exit code is the only remaining
  // signal. Without this gate, unrecognized JSON (including a bare `{}`)
  // silently decoded to allow regardless of exit code -- this is what
  // decodeKnownShape's "recognized" flag now prevents, mirroring the
  // claude/codex codecs. Gemini's wire never reveals which event a
  // decision answers, so decision.event is always left unset
  // (invariant 6).
  decodeDecision(stdout: unknown, exit: number): Decision {
    if (isEmptyStdout(stdout)) {
      return { action: actionForExit(this.caps, exit) };
    }
    const m = asRecord(stdout, "gemini decision");
    const { decision, recognized } = decodeKnownShape(m);
    if (!recognized) {
      return { action: actionForExit(this.caps, exit) };
    }
    return decision;
  }

  // checkEvent rejects a decision naming an event Gemini has no wire
  // shape for. A blank event keeps the default shape.
  private checkEvent(ev: Event | undefined): void {
    if (ev === undefined || ev === "") return;
    const { level: lvl, ok } = level(this.caps, ev);
    if (!ok || (lvl !== "native" && lvl !== "close")) {
      throw new UnsupportedEventError("gemini", ev);
    }
  }
}

function requireNativeOrClose(caps: Capabilities, host: string, event: Event): void {
  const { level: lvl, ok } = level(caps, event);
  if (!ok || (lvl !== "native" && lvl !== "close")) {
    throw new UnsupportedEventError(host, event);
  }
}

// decodeKnownShape probes the wire object for Gemini's known decision
// shapes and reports whether any were found. The systemMessage-based
// warn/allow fold is preserved exactly as before this gate was added: it
// disambiguates two shapes already recognized via "decision" rather than
// participating in recognition itself, mirroring hooks/hosts/gemini/codec.go.
function decodeKnownShape(m: Record<string, unknown>): { decision: Decision; recognized: boolean } {
  const d: Decision = { action: "allow" };
  let recognized = false;

  if (m.decision === "deny") {
    recognized = true;
    d.action = "block";
  } else if (m.decision === "allow") {
    recognized = true;
    d.action = "allow";
  }
  d.message = typeof m.reason === "string" ? m.reason : "";
  // The wire is lossy: plain allow and warn both carry decision:"allow"
  // plus "reason". Only "systemMessage" (set exclusively by the Warn
  // branch) distinguishes them; its absence means this genuinely cannot
  // be told apart from allow.
  if (typeof m.systemMessage === "string" && m.systemMessage !== "") {
    d.action = "warn";
  }
  const hso = isRecord(m.hookSpecificOutput) ? m.hookSpecificOutput : undefined;
  if (hso && "tool_input" in hso) {
    recognized = true;
    d.action = "rewrite";
    d.rewrite = hso.tool_input;
  }
  return { decision: d, recognized };
}

// decodeMap: native rows always win over close rows. Gemini's own
// motivating case: host event SessionEnd is listed native -> SessionEnd
// and close -> Stop; only the native row is what Gemini actually emits.
function decodeMap(caps: Capabilities): Map<string, Event> {
  const m = new Map<string, Event>();
  for (const p of caps.native) {
    if (m.has(p.hostEvent)) {
      throw new SchemaError(`${caps.host}: host event ${p.hostEvent} maps to both ${m.get(p.hostEvent)} and ${p.event} under native`);
    }
    m.set(p.hostEvent, p.event);
  }
  const seenClose = new Map<string, Event>();
  for (const p of caps.close ?? []) {
    if (seenClose.has(p.hostEvent)) {
      throw new SchemaError(`${caps.host}: host event ${p.hostEvent} maps to both ${seenClose.get(p.hostEvent)} and ${p.event} under close`);
    }
    seenClose.set(p.hostEvent, p.event);
    if (m.has(p.hostEvent)) continue;
    m.set(p.hostEvent, p.event);
  }
  return m;
}

function actionForExit(caps: Capabilities, exit: number): Decision["action"] {
  const codes = hostExitCodes(caps);
  if (exit === codes.block && exit !== codes.allow) return "block";
  return "allow";
}

function hostExitCodes(caps: Capabilities): { allow: number; block: number; warn?: number; rewrite?: number } {
  const h = requireHost(caps.host);
  if (!h.exitCodes) throw new SchemaError(`${caps.host}: no exit_codes in host.yaml`);
  return h.exitCodes;
}

function requireHost(name: string) {
  const h = get(name);
  if (!h) throw new SchemaError(`host ${name} not found`);
  return h;
}

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function asRecord(v: unknown, what: string): Record<string, unknown> {
  if (!isRecord(v)) throw new SchemaError(`${what}: expected a JSON object`);
  return v;
}

function isEmptyStdout(stdout: unknown): boolean {
  if (stdout === undefined || stdout === null || stdout === "") return true;
  if (typeof stdout === "string") return stdout.trim() === "";
  return false;
}

export function makeCodec(): Codec {
  return new GeminiCodec();
}

File: py/src/axon/hooks/hosts/gemini.py
"""The Gemini CLI codec. Mirrors hooks/hosts/gemini/codec.go (post the
T-0026 fail-open fix: DecodeDecision now has the same recognized-shape
gate as claude and codex); matched against ts/src/hooks/hosts/gemini.ts
for shape.
"""

from __future__ import annotations

from axon.capabilities import Capabilities, host_event as cap_host_event, level, load_capabilities
from axon.errors import SchemaError, UnsupportedActionError, UnsupportedEventError
from axon.hooks.decision import Decision
from axon.hooks.input import Input
from axon.registry import get

_KNOWN_INPUT_KEYS = {"hook_event_name", "session_id", "cwd", "tool_name", "tool_input", "tool_response", "prompt"}


class GeminiCodec:
    def __init__(self) -> None:
        self._caps = load_capabilities("gemini")
        self._host_to_canonical = _decode_map(self._caps)

    def host(self) -> str:
        return "gemini"

    def capabilities(self) -> Capabilities:
        return self._caps

    # encode_input fails with UnsupportedEventError when the event has no
    # native or close mapping: synthesized, unsupported, or unclassified
    # events cannot be encoded as a Gemini-native payload since Gemini
    # itself never emits them.
    def encode_input(self, inp: Input) -> dict:
        _require_native_or_close(self._caps, "gemini", inp.event)
        host_event_name = cap_host_event(self._caps, inp.event)
        if host_event_name is None:
            raise UnsupportedEventError("gemini", inp.event)

        out: dict = dict(inp.extra or {})
        out["hook_event_name"] = host_event_name
        out["session_id"] = inp.session_id
        out["cwd"] = inp.cwd
        if inp.event == "PreToolUse":
            out["tool_name"] = inp.tool_name
            out["tool_input"] = inp.tool_input
        elif inp.event == "PostToolUse":
            out["tool_name"] = inp.tool_name
            out["tool_input"] = inp.tool_input
            out["tool_response"] = inp.tool_response
        elif inp.event == "UserPromptSubmit":
            out["prompt"] = inp.prompt
        return out

    # decode_input looks up the canonical event in the native-wins decode
    # map on hook_event_name; a name not there is UnsupportedEventError,
    # never a silent fallback.
    def decode_input(self, raw: dict) -> Input:
        m = _as_dict(raw, "gemini input")
        host_event_name = m.get("hook_event_name")
        if not isinstance(host_event_name, str) or host_event_name == "":
            raise SchemaError("gemini input: missing hook_event_name")
        event = self._host_to_canonical.get(host_event_name)
        if event is None:
            raise UnsupportedEventError("gemini", host_event_name)

        extra = {k: v for k, v in m.items() if k not in _KNOWN_INPUT_KEYS}

        inp = Input(
            event=event,
            session_id=m.get("session_id") if isinstance(m.get("session_id"), str) else "",
            cwd=m.get("cwd") if isinstance(m.get("cwd"), str) else "",
            extra=extra,
        )
        if isinstance(m.get("tool_name"), str):
            inp.tool_name = m["tool_name"]
        if isinstance(m.get("tool_input"), dict):
            inp.tool_input = m["tool_input"]
        if isinstance(m.get("tool_response"), dict):
            inp.tool_response = m["tool_response"]
        if isinstance(m.get("prompt"), str):
            inp.prompt = m["prompt"]
        return inp

    # encode_decision emits Gemini's native decision envelope. Gemini has
    # no native "warn"; nerv folds warn into decision "allow" and, when
    # the message is non-empty, carries it in BOTH "reason" and
    # "systemMessage" (invariant 2: never mapped to block). Gemini does
    # not vary its wire shape by event (invariant 6): the event gate
    # below only rejects an event Gemini cannot express at all; it does
    # not change the shape.
    def encode_decision(self, d: Decision) -> tuple[object, int]:
        self._check_event(d.event)
        codes = _host_exit_codes(self._caps)
        if d.action == "allow":
            return {"decision": "allow"}, codes.allow
        if d.action == "warn":
            out: dict = {"decision": "allow"}
            if d.message:
                out["reason"] = d.message
                out["systemMessage"] = d.message
            # host.yaml carries no warn exit code for gemini; warn exits
            # like allow.
            return out, codes.allow
        if d.action == "rewrite":
            # host.yaml carries no rewrite exit code for gemini; rewrite
            # still allows the tool call to proceed, so it exits like
            # allow.
            return {"decision": "allow", "hookSpecificOutput": {"tool_input": d.rewrite}}, codes.allow
        if d.action == "block":
            return {"decision": "deny", "reason": d.message or ""}, codes.block
        raise UnsupportedActionError(f"gemini cannot express action {d.action!r}")

    # decode_decision reads Gemini's native decision envelope. Empty
    # stdout, or JSON that carries none of Gemini's known decision keys,
    # both fall back to _action_for_exit(exit): the exit code is the
    # only remaining signal. Gemini's wire never reveals which event a
    # decision answers, so decision.event is always left blank
    # (invariant 6).
    def decode_decision(self, stdout: object, exit_code: int) -> Decision:
        if _is_empty_stdout(stdout):
            return Decision(action=_action_for_exit(self._caps, exit_code))
        m = _as_dict(stdout, "gemini decision")
        decision, recognized = _decode_known_shape(m)
        if not recognized:
            return Decision(action=_action_for_exit(self._caps, exit_code))
        return decision

    # _check_event rejects a decision naming an event Gemini has no wire
    # shape for. A blank event keeps the default shape.
    def _check_event(self, ev: str) -> None:
        if not ev:
            return
        lvl, ok = level(self._caps, ev)
        if not ok or lvl not in ("native", "close"):
            raise UnsupportedEventError("gemini", ev)


def _require_native_or_close(caps: Capabilities, host: str, event: str) -> None:
    lvl, ok = level(caps, event)
    if not ok or lvl not in ("native", "close"):
        raise UnsupportedEventError(host, event)


# _decode_known_shape probes the wire object for Gemini's known decision
# shapes and reports whether any were found. The systemMessage-based
# warn/allow fold is preserved exactly: it disambiguates two shapes
# already recognized via "decision" rather than participating in
# recognition itself, mirroring hooks/hosts/gemini/codec.go.
def _decode_known_shape(m: dict) -> tuple[Decision, bool]:
    d = Decision(action="allow")
    recognized = False

    decision_val = m.get("decision")
    if decision_val == "deny":
        recognized = True
        d.action = "block"
    elif decision_val == "allow":
        recognized = True
        d.action = "allow"
    reason = m.get("reason")
    d.message = reason if isinstance(reason, str) else ""
    # The wire is lossy: plain allow and warn both carry
    # decision:"allow" plus "reason". Only "systemMessage" (set
    # exclusively by the Warn branch) distinguishes them; its absence
    # means this genuinely cannot be told apart from allow.
    system_message = m.get("systemMessage")
    if isinstance(system_message, str) and system_message != "":
        d.action = "warn"
    hso = m.get("hookSpecificOutput") if isinstance(m.get("hookSpecificOutput"), dict) else None
    if hso is not None and "tool_input" in hso:
        recognized = True
        d.action = "rewrite"
        d.rewrite = hso["tool_input"]
    return d, recognized


# _decode_map: native rows always win over close rows. Gemini's own
# motivating case: host event SessionEnd is listed native -> SessionEnd
# and close -> Stop; only the native row is what Gemini actually emits.
def _decode_map(caps: Capabilities) -> dict[str, str]:
    m: dict[str, str] = {}
    for p in caps.native:
        if p.host_event in m:
            raise SchemaError(
                f"{caps.host}: host event {p.host_event} maps to both {m[p.host_event]} and {p.event} under native"
            )
        m[p.host_event] = p.event
    seen_close: dict[str, str] = {}
    for p in caps.close or []:
        if p.host_event in seen_close:
            raise SchemaError(
                f"{caps.host}: host event {p.host_event} maps to both "
                f"{seen_close[p.host_event]} and {p.event} under close"
            )
        seen_close[p.host_event] = p.event
        if p.host_event in m:
            continue
        m[p.host_event] = p.event
    return m


def _action_for_exit(caps: Capabilities, exit_code: int) -> str:
    codes = _host_exit_codes(caps)
    if exit_code == codes.block and exit_code != codes.allow:
        return "block"
    return "allow"


def _host_exit_codes(caps: Capabilities):
    h = _require_host(caps.host)
    if h.exit_codes is None:
        raise SchemaError(f"{caps.host}: no exit_codes in host.yaml")
    return h.exit_codes


def _require_host(name: str):
    h = get(name)
    if h is None:
        raise SchemaError(f"host {name} not found")
    return h


def _as_dict(v: object, what: str) -> dict:
    if not isinstance(v, dict):
        raise SchemaError(f"{what}: expected a JSON object")
    return v


def _is_empty_stdout(stdout: object) -> bool:
    if stdout is None or stdout == "":
        return True
    if isinstance(stdout, str):
        return stdout.strip() == ""
    return False


def make_codec() -> GeminiCodec:
    return GeminiCodec()
